timemachine/optimize/protocol.py

```diff
@@ -208,6 +219,56 @@ def max_work_stddev_distance(prev_lam, next_lam, max_step=0.25):
     return max_work_stddev_distance
 
 
+# (2) pair overlap
+def reweighted_pair_overlap(u_n_A, u_n_B, u_n_ref):
+    """given arrays [u(x) for x in xs] (for u in {u_A, u_B, u_ref}),
+    estimate overlap(A, B) by reweighting from ref->A and ref->B"""
+    # reduced potentials -> unnormalized log probs
+    log_q_A = -u_n_A
+    log_q_B = -u_n_B
+    log_q_ref = -u_n_ref
+    log_N = jnp.log(len(log_q_ref))
+
+    # pick a normalization constant for ref
+    log_p_ref_n = log_q_ref - logsumexp(log_q_ref - log_N)
+
+    # normalize A and B wrt ref
+    log_p_A = log_q_A - logsumexp(log_q_A - log_p_ref_n - log_N)
+    log_p_B = log_q_B - logsumexp(log_q_B - log_p_ref_n - log_N)
+
+    log_prod_AB_n = log_p_A + log_p_B
+    # assert (log_prod_AB_n < jnp.inf).all()
+
+    log_p_mix_n = logsumexp(jnp.array([log_p_A, log_p_B]), axis=0)
+    # assert (log_p_mix_n < jnp.inf).all()
+
+    log_denom = log_p_mix_n + log_p_ref_n
+    mask = log_denom > -jnp.inf  # mask out div by 0
+    log_f_n = jnp.where(mask, log_prod_AB_n - log_denom, 0.0)
+    # assert not jnp.isnan(log_f_n).any()
+
+    log_mean_f = logsumexp(log_f_n - jnp.log(sum(mask)))
+    overlap = jnp.exp(log_mean_f)
+    return 2 * overlap  # so that it goes to 1.0 when u_n_A == u_n_B
+
+
+def make_approx_overlap_distance_fxn(lambdas, u_kn, f_k, N_k):
+    """make a distance function d(a,b) = 1 - overlap(a,b)
+
+    where overlap(a,b) uses fast approximations: reweighting, based on linear interpolation of energies"""
+    u_lam = linear_u_kn_interpolant(lambdas, np.nan_to_num(u_kn, nan=np.inf))
```

Comment:
On the fence: might want to make this optimization optional

Reply:
What are the situations in which this would break down?

Reply:
Reading the docstrings:

```
Given a matrix u_kn[k, n] = u(xs[n], lambdas[k]) produce linear interpolated estimates of u(xs[n], lam) at arbitrary new values lam
```

Our 4D nonbonded softcore implementation and log-linear bond interpolation schemes would probably result in very non-linear energies as a function of lambda.

Reply:
> This assumption can be reasonable -- even though u is a nonlinear function of lam -- if lambdas[k] is dense enough.

Is there a rule of thumb/ballpark estimate on what "dense enough" and "reasonable" is? I don't know if the initial bisection protocol makes guarantees about amount of error introduced by a linear approximation of the energy. (I might have missed these values originally in #538). 

Can we quantify this error in some way? Some ideas that come to mind are:

For each [solvent, vacuum] using `st`:
1) [unit test] Assert on the maximum energy deviation relative to the reference energies over multiple trajectories (may be too stringent).
2) [unit test] Assert on the `reweighted_pair_overlap` using the linear approximation energy vs the reference energy
3) [integration test] See if we end up with approximately the same # of windows after greedily optimizing the protocol. i.e. do we know that the reduction in # of windows isn't a result of the linear approximation?

Reply:
These are good test ideas. This is an uncontrolled approximation, and I only spot-checked approximation quality in a cursory way.

The spot-checks I've done so far are most similar to your suggestion (2) -- compare reweighting estimates based on approximate energies against reweighting estimates based on reference energies.

Here's an example of a comparison from internal documentation, using the `get_hif2a_ligand_pair_single_topology` testsystem* (*modified: 10 atoms truncated from the atom mapping).

![image](https://github.com/user-attachments/assets/3013cba3-6cbb-4ee0-806b-2744ccff2cc7)

x-axis is `lam`, y-axis is estimated `overlap(lam, 0.25)`

blue dots are from reweighting using linear interpolation of `u_kn`, red dots are from reweighting using the potential energy function.

Will expand this sort of comparison into a test...

> I don't know if the initial bisection protocol makes guarantees about amount of error introduced by a linear approximation of the energy.

I don't think it does.

Reply:
> Here's an example of a comparison from internal documentation, using the `get_hif2a_ligand_pair_single_topology` testsystem* (*modified: 10 atoms truncated from the atom mapping).
> 
> ![image](https://private-user-images.githubusercontent.com/5759036/374688505-3013cba3-6cbb-4ee0-806b-2744ccff2cc7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0MTQ5MDIsIm5iZiI6MTcyODQxNDYwMiwicGF0aCI6Ii81NzU5MDM2LzM3NDY4ODUwNS0zMDEzY2JhMy02Y2JiLTRlZTAtODA2Yi0yNzQ0Y2NmZjJjYzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDhUMTkxMDAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzcwODQyOGUyZWJkN2VjNTQ0MTU1ZDJjNDUyYWU2YjczZDNiMDcxNWFlZWIyMjZkZWQwODFkODBiNGJhOTIwYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GTQFfgCvtKxxDI6b12HhFBOIdAIhk_JJsUXNRdWT4l4)
> 
> x-axis is `lam`, y-axis is estimated `overlap(lam, 0.25)`
> 
> blue dots are from reweighting using linear interpolation of `u_kn`, red dots are from reweighting using the potential energy function.

Nice! This plot looks re-assuring to me. Clarification: was this done in vacuum? If the plots look similar in solvent/complex it would significantly boost my confidence in this approximation (rough thinking is that the # of interactions and sensitivity of energies to lambda is greater when nonbonded interaction groups are present). 

> Will expand this sort of comparison into a test...

Awesome!

Reply:
>Clarification: was this done in vacuum? If the plots look similar in solvent/complex it would significantly boost my confidence in this approximation (rough thinking is that the # of interactions and sensitivity of energies to lambda is greater when nonbonded interaction groups are present).

This one's vacuum : will expand.

Reply:
Running a larger spot-check in vacuum with a large number of random `(lam_a, lam_b)` pairs, and comparing the estimated overlap using reference potential energy function vs. linearly interpolated potential energy fxn produces roughly the following comparison:

![image](https://github.com/user-attachments/assets/543f2432-9719-4382-9440-6d2b1a3e0047)

close agreement most of the time (MAE ~= 0.00841), but occasional large disagreement (max disagreement found so far ~= 0.256). also, occasionally out of expected bounds.

Reply:
Sorry for missing detail.

> Are (lam_a, lam_b) always adjacent pairs from the bisected schedule?

No: linear interpolation is exact cases where `lam_a, lam_b` are both in `set(lambdas)`, and there are only a few dozen of those discrete adjacent pairs to check. 

In the above plot, 1000 random `(lam_a, lam_b)` pairs were sampled according to: `lam_a = 0.8 * rand(); lam_b = lam_a + 0.2 * rand()`. Will be curious to figure out what's going on in the cases with the largest estimation error.

Reply:
Thanks for the additional clarification.

Perhaps intentional but the scheme you proposed uniformly samples `lam_a in [0, 0.8]`, but does *not* uniformly sample `lam_b` (ie. there will be very few values of `lam_b > 0.95` tested relative to `lam_a < 0.05`). i.e. this ultimately may not stress test the higher values of lambda as much.

![image](https://github.com/user-attachments/assets/c9aa0356-43bb-4b20-96e1-c274a946a3bb)

Some additional empirical observations of the error plot:

1. In the cases where error is significant (>0.1), the estimated overlap based on interpolated energies seems to consistently greater than the reference energies. If it were the other around that's less of a concern given the use case: i.e. (underestimating the overlap is a more defensive outcome than overestimating the overlap if the ultimate goal is to rebalance the number of windows).

2. There appears to be two additional off-diagonal bands in the upper right quadrant of the plot separated by 0.1 apart. 

3. As you noted, not sure how to interpret out of bounds where `estimated_pair_overlap > 1`.

Reply:
> Perhaps intentional but the scheme you proposed uniformly samples lam_a in [0, 0.8], but does not uniformly sample lam_b

That's true: could adjust the sampling distribution of this spot-check.

(In the spot-check, I sampled a uniform random interval width `width ~ Unif(0, 0.2)`, and a uniform random left bound `lam_a ~ Unif(0, 0.8)`. Could instead do something like sample a uniform `midpoint ~ Unif(width/2, 1 - width/2)`, `lam_a = midpoint - width/2, lam_b = midpoint + width/2`, to be more symmetric.)

Reply:
Investigation is still pending. One additional note: even if the approximation quality is poor for some query points `(lam_a, lam_b)`, the overall context of the optimizer could still be minimally impacted. (i.e. the sequence of decisions made by the optimizer might be about the same with or without this approximation).

Note that the optimizer interacts with this approximation **only by checking the sign** of `overlap(lam_a, lam_b) - target_val`, where `target_val` is maybe 0.1 or 0.25 or 0.5 or 0.666. So, even if this approximation introduces a large absolute error on some query points (e.g. `approx_overlap(a, b) = 1.2` when `overlap(a, b) = 0.95`), the error might be inconsequential to the optimizer.

For illustration, here is the sequence of estimates produced on the sequence of points queried by the optimizer, with a target overlap of 0.25 between adjacent states:

![image](https://github.com/user-attachments/assets/776ef61f-8809-45a4-91ec-19e71593bd63)

Here is a plot of the agreement between the approximation and the reference on this sequence of points:

![image](https://github.com/user-attachments/assets/59aaedf6-be8f-4f98-8200-557bbee58bdf)

The approximation gets the sign right on ~90% of the points queried in this particular sequence. Although there are queried points where the approximation is very wrong in terms of absolute error, the approximation gets the sign right on those points (i.e. the large errors do not lead to a different optimizer outcome). In the cases where it gets the sign wrong, the absolute approximation error seems small (say, < 0.015).

Still want to figure out what's causing the estimate to sometimes exceed 1 though...

... Anyway, if the approach looks unrecoverable after further investigation, a different approach that might make sense is to try to approximate the whole function `overlap(a,b) -> R`, rather than trying to approximate the function `u(lam) -> R^N` that appears as an intermediate step of the reweighting approximation.

Reply:
> Will be curious to figure out what's going on in the cases with the largest estimation error.

Looking at this again: appears to be due to `f64` -> `f32` precision loss in the step that linearly interpolates `u_kn` (which starts in f64)

> linear interpolation is exact in cases where lam_a, lam_b are both in set(lambdas)

This statement was not true, due to precision loss